jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/paths/PathCollection.java

```diff
@@ -0,0 +1,291 @@
+            FileSystemPool.release(old);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException("Unable to release old path: " + old);
+        }
+        return old;
+    }
+
+    @Override
+    public void clear()
+    {
+        RuntimeException error = null;
+
+        for (Path path : this)
+        {
+            try
+            {
+                FileSystemPool.release(path);
+            }
+            catch (IOException e)
+            {
+                if (error == null)
+                    error = new RuntimeException("Unable to release some paths");
+                error.addSuppressed(e);
+            }
+        }
+        if (error != null)
+            throw error;
+    }
+
+    /**
+     * Resolve the first path that exists against the collection of paths.
+     *
+     * @param other the other path to resolve against, assumes a String in unix format, that is already URI decoded from URI space.
+     * @see Path#resolve(String)
+     */
+    public Path resolveFirstExisting(String other)
+    {
+        return resolveFirst(other, Files::exists);
+    }
+
+    /**
+     * Resolve the first path that matches the predicate.
+     *
+     * @param other the other path to resolve against, assumes a String in unix format, that is already URI decoded from URI space.
+     * @param pathPredicate the predicate of {@code Path} to evaluate for first valid hit.
+     */
+    public Path resolveFirst(String other, Predicate<Path> pathPredicate)
+    {
+        for (Path path : this)
+        {
+            Path dest = path.resolve(other);
+            if (dest.startsWith(path) && pathPredicate.test(dest))
+                return dest;
+        }
+        return null;
+    }
+
+    /**
+     * Resolve all paths across the path collection that matches the provided other.
+     *
+     * @param other the other path to resolve against, assumes a String in unix format, that is already URI decoded from URI space.
+     * @param pathPredicate the predicate of {@code Path} to evaluate for valid hit.
+     * @return the list of {@link Path} objects that satisfy the combination of {@code other} and {@code Predicate<Path>}
+     */
+    public List<Path> resolveAll(String other, Predicate<Path> pathPredicate)
+    {
+        List<Path> ret = new ArrayList<>();
+        for (Path path : this)
+        {
+            Path dest = path.resolve(other);
+            if (dest.startsWith(path) && pathPredicate.test(dest))
+                ret.add(dest);
+        }
+        return ret;
+    }
+
+    /**
+     * Find all files (up to 300 levels deep) in the path collection that match the provided {@link BiPredicate}
+     * of &lt;{@link Path}, {@link BasicFileAttributes}&gt;
+     *
+     * @param pathPredicate the predicate to evaluate across all paths in this collection
+     * @return the stream of hits for the find operation.
+     */
+    public Stream<Path> find(BiPredicate<Path, BasicFileAttributes> pathPredicate)
```

Comment:
you may as well pass in the depth rather than hard code to 300.  This makes it very clear the difference with resolveAll, which is essentially a find with depth 0